pkg/plugin/eventqueueset.go

```diff
@@ -0,0 +1,56 @@
+package plugin
+
+import (
+	"context"
+	"hash/fnv"
+	"time"
+
+	"github.com/tychoish/fun/pubsub"
+)
+
+type queueItem[T any] struct {
+	item    T
+	addTime time.Time
+}
+
+type eventQueueSet[T any] struct {
+	queues  []*pubsub.Queue[queueItem[T]]
+	metrics PromMetrics
+}
+
+func newEventQueueSet[T any](size int, metrics PromMetrics) eventQueueSet[T] {
+	queues := make([]*pubsub.Queue[queueItem[T]], size)
+	for i := 0; i < size; i += 1 {
+		queues[i] = pubsub.NewUnlimitedQueue[queueItem[T]]()
+	}
+	return eventQueueSet[T]{
+		queues:  queues,
+		metrics: metrics,
+	}
+}
+
+func (s eventQueueSet[T]) enqueue(key string, item T) error {
+	hasher := fnv.New64()
+	// nb: Hash guarantees that Write never returns an error
+	_, _ = hasher.Write([]byte(key))
+	hash := hasher.Sum64()
+
+	idx := int(hash % uint64(len(s.queues)))
+
+	s.metrics.eventQueueDepth.Inc()
+	s.metrics.eventQueueAddsTotal.Inc()
+	queueItem := queueItem[T]{
+		item:    item,
+		addTime: time.Now(),
+	}
+	return s.queues[idx].Add(queueItem)
+}
+
+func (s eventQueueSet[T]) wait(ctx context.Context, idx int) (T, error) {
+	queueItem, err := s.queues[idx].Wait(ctx)
+
+	s.metrics.eventQueueDepth.Dec()
+	s.metrics.eventQueueLatency.Observe(float64(time.Since(queueItem.addTime).Seconds()))
```

Comment:
This is only valid if err != nil

Reply:
Right. I wanted to capture the latency even if we got an error, but I forget we don't get the right item in this case. Fixed in 129dd3529d0c53c0531a8b9f84020525e8befe47.

Reply:
Same thing with decreasing the event queue depth, I think?

Reply:
Fixed.